src/Algebra/Graph.hs

```diff
@@ -1202,81 +1310,63 @@ sparsifyKL n graph = KL.buildG (1, next - 1) ((n + 1, n + 2) : Exts.toList (res
 The rules for foldg work very similarly to GHC's mapFB rules; see a note below
 this line: http://hackage.haskell.org/package/base/docs/src/GHC.Base.html#mapFB.
 
-* Up to (but not including) phase 1, we use the "buildR/f" rule to rewrite all
-  saturated applications of f into its buildR/foldg form, hoping for fusion to
-  happen (through the "foldg/buildR" rule).
-
-  In phases 1 and 0, we switch off these rules, inline buildR, and switch on the
-  "graph/f" rule, which rewrites "foldg/f" back into plain functions if needed.
+* All concerned expressions are inlined to allow the compiler to apply the main
+  rule: "foldg/buildg".
+  This rule states that the composition of a good producer (expressed via buildg)
+  and a good consumer (expressed via foldg) can be fused to remove the construction
+  of the intermediate structure.
 
-  It's important that these two rules aren't both active at once (along with
-  build's unfolding) else we'd get an infinite loop in the rules. Hence the
-  activation control below.
+* If this inlining is made blindlessly, it can lead to unneeded operations. They
+  are optimized via the "foldg/id" rule.
 
-* composeR and matchR are here to remember the original function after applying
-  a "buildR/f" rule. These functions are higher-order functions and therefore
+* composeR is here to allow further optimization. As an high-order function, it
   benefit from inlining in the final phase.
 
-* The "bindR/bindR" rule optimises compositions of multiple bindR's.
+* The "composeR/composeR" rule optimises compositions of multiple composeR's.
 -}
 
-type Foldg a = forall b. b -> (a -> b) -> (b -> b -> b) -> (b -> b -> b) -> b
-
-buildR :: Foldg a -> Graph a
-buildR g = g Empty Vertex Overlay Connect
-{-# INLINE [1] buildR #-}
+-- | Build a graph given an interpretation of the four graph construction primitives 'empty',
```

Comment:
Let's move this definition up, after `foldg`, to match the export order (just for ease of navigation).